### DISCUSSION TITLE
Opening and make use of several midi-in-devices at once?

### DISCUSSION THREAD

**Discussion by rolfiepolfie** [mido/mido#433] [Q&A] [answered]
I just wondered  if this possible or advised in any form reading MIDI-in data from multiple devices at the same time?
 
Reading about virtual ports in the manual and wondering if that can help out. Anyway, I got two different devices wanting to read midi data from both (plugged into the USB ports, two midi-controllers with a usb only)

 I must either:
1. open them both in code and use at the same time 
2. use one device only  and merge several external MIDI signals into it by other methods 

I see hardware midi-mergers around, as: https://www.thomannmusic.no/midi_solutions_merger_v2.htm?gclid=Cj0KCQjwu8r4BRCzARIsAA21i_AeHuaX0bUle7tmmdlvp6gTDpeKY67VP2hB8LYtM2RYo2ujq1dN2X4aAh9oEALw_wcB

.... but my stuff only got USB connection, not old-type MIDI connections pictured. Do I have to do a deep operations in the backend or is there a simpler solution, I wonder how to attack this!!! Thanks

**Comment by olemb** [comment, marked as answer]
Just open each device with `mido.open_input()` and use them at the same time. They will not interfere with each other and each port will receive messages from the correct device.

With the default backend (python-rtmidi) you can even open multiple ports to the same device and the messages will be delivered on all ports. This can be useful sometimes if you want to handle the same messages in different parts of the code, or by different processes or threads.

If you want to wait for messages from both ports at the same time check out `mido.ports.multi_receive()`. (Or you can set up the ports to both use the same callback.)

**Comment by rolfiepolfie** [comment]
somewhat delayed, but thankyou for the reply, I have resumed my MIDI progging , and will continue until my official death
